src/main.rs

```diff
@@ -495,3 +512,38 @@ fn main_recv(addr: &str, n_conn: &str) -> Result<()> {
 
     Ok(())
 }
+
+#[cfg(test)]
```

Comment:
Thanks for adding tests!

For the approach though … testing something involving IO like this is always tricky. As you say if we hard-code the port it’s annoying if you already have something listening there (which could be a previous instance of the test, which tends to happen in tests when the program fails and the socket remains in `TIME_WAIT` …). Also, putting `thread::sleep` in tests makes the tests slow and often flaky.

The “proper” way to test it I think would be to separate the logic and the IO, and then we can test the logic without any binds or sleeps. (But then the test is also not testing the real thing any more.) For a large application that worth it, but for a tiny program like this, I think it would double the amount of code.

So I’m not sure what to do in this case, it’s probably better than no test at all and splitting out the logic and IO would be overkill. Let’s keep it then :+1:

Reply:
thanks to the suggestion to use channels for synchronization, i'm now able to use a free port (`0`).